rust-runtime/aws-smithy-http-server-python/src/middleware/layer.rs

```diff
@@ -50,175 +50,79 @@ where
-    use crate::{PyMiddlewareException, PyMiddlewareType, PyRequest};
-
-    async fn echo(req: Request<Body>) -> Result<Response<BoxBody>, Box<dyn Error + Send + Sync>> {
-        Ok(Response::new(to_boxed(req.into_body())))
-    }
-
-    #[tokio::test]
-    async fn request_middlewares_are_chained_inside_layer() -> PyResult<()> {
-        let locals = crate::tests::initialize();
-        let mut middlewares = PyMiddlewares::new::<RestJson1>(vec![]);
-
-        Python::with_gil(|py| {
-            let middleware = PyModule::new(py, "middleware").unwrap();
-            middleware.add_class::<PyRequest>().unwrap();
-            middleware.add_class::<PyMiddlewareException>().unwrap();
-            let pycode = r#"
-def first_middleware(request: Request):
-    request.set_header("x-amzn-answer", "42")
-    return request
-
-def second_middleware(request: Request):
-    if request.get_header("x-amzn-answer") != "42":
-        raise MiddlewareException("wrong answer", 401)
-"#;
-            py.run(pycode, Some(middleware.dict()), None)?;
-            let all = middleware.index()?;
-            let first_middleware = PyMiddlewareHandler {
-                func: middleware.getattr("first_middleware")?.into_py(py),
-                is_coroutine: false,
-                name: "first".to_string(),
-                _type: PyMiddlewareType::Request,
-            };
-            all.append("first_middleware")?;
-            middlewares.push(first_middleware);
-            let second_middleware = PyMiddlewareHandler {
-                func: middleware.getattr("second_middleware")?.into_py(py),
-                is_coroutine: false,
-                name: "second".to_string(),
-                _type: PyMiddlewareType::Request,
-            };
-            all.append("second_middleware")?;
-            middlewares.push(second_middleware);
-            Ok::<(), PyErr>(())
-        })?;
-
-        let mut service = ServiceBuilder::new()
-            .layer(PyMiddlewareLayer::<RestJson1>::new(middlewares, locals))
-            .service_fn(echo);
-
-        let request = Request::get("/").body(Body::empty()).unwrap();
-
-        let res = service.ready().await.unwrap().call(request).await.unwrap();
-
-        assert_eq!(res.status(), 200);
-        Ok(())
+        let inner = {
+            // https://docs.rs/tower/latest/tower/trait.Service.html#be-careful-when-cloning-inner-services
+            let clone = self.inner.clone();
+            std::mem::replace(&mut self.inner, clone)
+        };
+        let handler = self.handler.clone();
+        let handler_name = handler.name.clone();
+        let next = BoxService::new(inner.map_err(|err| err.into()));
+        let locals = self.locals.clone();
+        let into_response = self.into_response;
+
+        Box::pin(
+            handler
+                .call(req, next, locals)
+                .or_else(move |err| async move {
+                    tracing::error!(err = %err, "middleware '{handler_name}' failed");
```

Comment:
By only logging the display representation of the error (`%err`) I don't think you are capturing the traceback information.

Reply:
Yes good point, I've added `RichPyErr` to add more detail to `Debug` implementation of `PyErr` https://github.com/awslabs/smithy-rs/commit/6db19b31a3d8da5b3e0b58337fb42ebd0def89a3